FLEDGE.md

```diff
@@ -281,6 +288,42 @@ Once all of a component auction's bids have been scored by the component auction
 The ultimate winner of the top-level auction is the single bid the top-level seller script gives the highest score. This may either be the winning bid of one of the component auctions, or a bid from one of the `interestGroupBuyers` in the `AuctionConfig` of the top-level auction. Those bids will be scored directly by the top-level seller script without having to win any component auction.
 
 
+#### 2.5 extraSignals Hidden from the Page (for Auction Worklets Only)
+
+Auction config parameter values are visible to any scripts, first or third party, that are running on the frame that called `runAdAuction()`. For instance, a script could assign a new proxy function to `navigator.runAdAuction()`.
+
+For auction fairness, sellers might not want to reveal seller and per-buyer signals to other parties.
+
+
+The optional `extraSellers` field can be used to pass signals to the auction, similar to `sellerSignals` and `perBuyerSignals`. The difference is that `extraSignals` content cannot be viewed by scripts running on the page -- the signals may only be read by the respective intended auction participants. If present, `extraSellers` should be an https URL of a resource in a [subresource bundle](https://github.com/WICG/webpackage/blob/main/explainers/subresource-loading.md) that has been loaded by the current document, whose contents are of type `application/json`, with the following response headers: `X-Allow-Fledge: true` and `X-FLEDGE-Auction-Only: true`.
+
+The JSON response should be a dictionary, whose keys are the origins of the auction participants (including both the seller and individual buyers), and whose values are the signals to be delivered to each respective auction participant.
+
+```
+{
+  'seller.com': {...},
+  'buyer1.com': {...},
+  'buyer2.com': {...}
```

Comment:
Should there be an `*` field?

Reply:
That seems like it would be useful, but this brings up a good point I forgot to mention -- if there are 2 keys, both for https://buyer1.com, we should fail loading, since we don't define a way to merge these signals.

'*' would inherently conflict with all other signals. However, we could make separate worklet parameters for this, similar to actionSignals.

Given the complexity this would add, for the first iteration of this, I'm going to focus on the base functionality, but this does seem like it would be useful in a subsequent iteration. I can mention that '*' is reserved (I don't think it's a valid origin, anyways).

Reply:
Note that `*` is the fallback, we'd be matching behavior elsewhere.

Also, I think we should do whatever our RFC compliant JSON parser does when there are duplicate values, rather than specifying it here, potentially requiring we add a special mode to our JSON parser if it doesn't match what behavior we specify?

Reply:
'*' as fallback seems reasonable, and easy to implement / test.

The RFC is permissive about duplicate keys; in the Chromium implementation, the [last duplicate silently wins](https://source.chromium.org/chromium/chromium/src/+/main:base/json/json_reader.h;l=20;drc=8e71b8677f674bab739791d555b1bc32df30c392).

Reply:
Another option to consider is to represent extra signals as an object with keys similar to those in the auction configuration, e.g.:
```jsonc
{
  "sellerSignals": {/*…*/},
  "auctionSignals": {/*…*/},
  "perBuyerSignals": {"https://buyer1.example": /*…*/}
}

Reply:
I like that idea - much more flexible format, though it does double-down on the issue of where we're going to parse these.

Reply:
@sbelov @MattMenke2 
What do you think of this proposal? 

Instead of a separate `extraAuctionSignals`, we could just have a single `extraSignals` field whose value is the URL prefix.

`?origin=[origin]` gets appended to the prefix. For `extraSellerSignals`, `origin` will be the seller's origin. For `extraPerBuyerSignals`, it will be an origin of one of the buyers. Finally, `?origin=*` indicates `extraAuctionSignals` (since everyone gets a copy of these) [0].

To me, this makes more sense than handling multiple prefix URLs for each of the 3 cases.

If this sounds good, I can go ahead and update the PR.

[0] `*` is allowed unencoded in URLs per https://www.ietf.org/rfc/rfc1738.txt

Reply:
@sbelov @MattMenke2 Friendly ping?

Reply:
I would rather have the seller signals and buyer signals distinct when an origin is both a buyer and a seller, but seems reasonable, though I'm not really sure how web packages work, or how they'd be mapped to a URL prefix.

Reply:
So, maybe instead of `?origin=[origin]` and `?origin=*`, we could have `?perBuyerSignals=[origin]`, `?sellerSignals=true`, and `?auctionSignals=true`. (A given subresource would only have one of these).

Maybe the `?` should be (a required) part of the prefix, to allow other URL query parameters to be set? 

> I'm not really sure how web packages work, or how they'd be mapped to a URL prefix.

Resources inside the subresource bundle are addressable by an https URL. When requesting this https URL, the network service first checks to see a bundle containing a subresource with that URL has already been registered (pages must declare the subresources as part of the <script type="application/webbundle"> tags -- this fixes races where the bundle may not be loaded when the subresource is requested.) If no bundle is found, the network service will then make a standard request over the network.

The query parameters will ensure that each signal will be a separate resource in the subresource bundle -- and the browser process will control the construction of the query parameters.

The bundle therefore wouldn't know / care about prefixes, it would just have several subresources that happen to share prefixes. Alternatively, the signals could be delivered across several bundle files.

Reply:
That works for me - we don't even need the "=true" - query params don't actually have to end in "=foo".

Reply:
The reason for this proposed feature is concern about publisher access of these fields in auctionConfig, and manipulation of the auctionConfig.  If these are in the auctionConfig, the publisher can potentially manipulate the URLs to switch data the various JS methods receive.

Admittedly, sellers and bidders could include information to protect against this in the dictionaries (e.g., bidder origin), though that would still allow publishers to pass information intended for one bidder to another, which I thin is something we might want to guard against?

Reply:
I've pushed a new commit that addresses the original issue of this comment, PTAL.